editors/code/src/tasks.ts

```diff
@@ -115,14 +109,26 @@ export async function buildCargoTask(
         exec = new vscode.ProcessExecution(fullCommand[0], fullCommand.slice(1), definition);
     }
 
-    return new vscode.Task(
-        definition,
-        target,
-        name,
-        TASK_SOURCE,
-        exec,
-        ['$rustc']
-    );
+    if (scope) {
+        return new vscode.Task(
+            definition,
+            scope,
+            name,
+            TASK_SOURCE,
+            exec,
+            ['$rustc']
+        );
+    }
+    else {
+        // if the original task did not provide a scope retain the original lack of scope
+        return new vscode.Task(
```

Comment:
This constructor is considered deprecated: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/vscode/index.d.ts#L6743

> * @deprecated Use the new constructors that allow specifying a scope for the task.

So looks like VSCode forces everybody to explicitly define a workspace folder for every task.

I'm not entirely thrilled to start using deprecated APIs, so would personally concentrate on deriving the folder instead.
Seems unfair to block you too, so if somebody else feels fine, he can merge the PR.

Reply:
Ah I see, so was your comment [here](https://github.com/rust-analyzer/rust-analyzer/pull/8995#issuecomment-909040741):
> Now, the biggest task is to determine the folder that should apply for your custom configuration since I did not take such cases into account.

about how to handle `scope === undefined`?

In that case I assumed that it should simply 'add on' the necessary parts to the task, and if the original task didn't have a scope then simply leave it undefined. From what I've seen, the function doesn't need to use the scope and so if it doesn't have one it just passes that lack of scope onwards. But once again, I've little understanding of what's going on here.

Alternatively could the check instead be `if (scope === undefined)`?

Reply:
You've got the idea right: it's all about the `scope === undefined` for tasks that come from user configurations (in your case).

The fix logic can be anything that fixes things for you and preferrably is not deprecated/super hacky 🙂 
I see that the scope can be `scope: WorkspaceFolder | TaskScope.Global | TaskScope.Workspace` so I wonder, if the `TaskScope.Workspace` variant a right thing to fix your issues?

Reply:
`TaskScope.Workspace` is indeed the scope that the tasks are given, the comments on `TaskScope.Global` seem to imply it's not used currently.